Source/WebCore/dom/Document.cpp

```diff
@@ -906,40 +911,46 @@ void Document::commonTeardown()
 
 Quirks& Document::ensureQuirks()
 {
+    ASSERT(m_constructionDidFinish);
```

Comment:
This ASSERT hits on the GTK Debug bots, notably when running inspector tests... cc @obyknovenius 

https://build.webkit.org/results/GTK-Linux-64-bit-Debug-Tests/273602@main%20(12275)/results.html

```
Thread 1 (Thread 0x7f85c817da40 (LWP 248487)):
#0  WTFCrash() () at /app/webkit/Source/WTF/wtf/Assertions.cpp:333
#1  0x00007f85de0bb486 in WTFCrashWithInfo(int, char const*, char const*, int) () at /app/webkit/WebKitBuild/GTK/Debug/WTF/Headers/wtf/Assertions.h:778
#2  0x00007f85e2e5b2d8 in WebCore::Document::ensureQuirks() (this=0x7f8576103000) at /app/webkit/Source/WebCore/dom/Document.cpp:917
#3  0x00007f85df800f3f in WebCore::Document::quirks() (this=0x7f8576103000) at /app/webkit/WebKitBuild/GTK/Debug/WebCore/PrivateHeaders/WebCore/DocumentInlines.h:60
#4  0x00007f85e30551c1 in WebCore::isTouchRelatedEventType(WebCore::EventTypeInfo const&, WebCore::EventTarget const&) (eventType=..., target=...) at /app/webkit/Source/WebCore/dom/Node.cpp:2359
#5  0x00007f85e305533f in WebCore::tryAddEventListener(WebCore::Node*, WTF::AtomString const&, WTF::Ref<WebCore::EventListener, WTF::RawPtrTraits<WebCore::EventListener> >&&, WebCore::AddEventListenerOptions const&) (targetNode=0x7f8576103000, eventType=..., listener=..., options=...) at /app/webkit/Source/WebCore/dom/Node.cpp:2379
#6  0x00007f85e3055405 in WebCore::Node::addEventListener(WTF::AtomString const&, WTF::Ref<WebCore::EventListener, WTF::RawPtrTraits<WebCore::EventListener> >&&, WebCore::AddEventListenerOptions const&) (this=0x7f8576103000, eventType=..., listener=..., options=...) at /app/webkit/Source/WebCore/dom/Node.cpp:2406
#7  0x00007f85e38f393d in operator()(WTF::AtomString const&) const (__closure=0x7ffcd6caa5b0, eventName=...) at /app/webkit/Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:2572
#8  0x00007f85e38f3a1c in WebCore::InspectorDOMAgent::addEventListenersToNode(WebCore::Node&) (this=0x7f85be1b3240, node=...) at /app/webkit/Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:2577
#9  0x00007f85e384c752 in WebCore::InspectorInstrumentation::addEventListenersToNodeImpl(WebCore::InstrumentingAgents&, WebCore::Node&) (instrumentingAgents=..., node=...) at /app/webkit/Source/WebCore/inspector/InspectorInstrumentation.cpp:144
#10 0x00007f85e2eac761 in WebCore::InspectorInstrumentation::addEventListenersToNode(WebCore::Node&) (node=...) at /app/webkit/Source/WebCore/inspector/InspectorInstrumentation.h:578
#11 0x00007f85e2e589b0 in WebCore::Document::Document(WebCore::LocalFrame*, WebCore::Settings const&, WTF::URL const&, WTF::OptionSet<WebCore::Document::DocumentClass>, WTF::OptionSet<WebCore::Document::ConstructionFlag>, WebCore::ProcessQualified<WTF::UUID>) (this=0x7f8576103000, frame=0x7f85be1001a0, settings=..., url=..., documentClasses=..., constructionFlags=..., identifier=...) at /app/webkit/Source/WebCore/dom/Document.cpp:650
#12 0x00007f85e33d01d5 in WebCore::HTMLDocument::HTMLDocument(WebCore::LocalFrame*, WebCore::Settings const&, WTF::URL const&, WebCore::ProcessQualified<WTF::UUID>, WTF::OptionSet<WebCore::Document::DocumentClass>, WTF::OptionSet<WebCore::Document::ConstructionFlag>) (this=0x7f8576103000, frame=0x7f85be1001a0, settings=..., url=..., documentIdentifier=..., documentClasses=..., constructionFlags=...) at /app/webkit/Source/WebCore/html/HTMLDocument.cpp:99
#13 0x00007f85e2e35b94 in WebCore::HTMLDocument::create(WebCore::LocalFrame*, WebCore::Settings const&, WTF::URL const&, WebCore::ProcessQualified<WTF::UUID>) (frame=0x7f85be1001a0, settings=..., url=..., identifier=...) at /app/webkit/Source/WebCore/html/HTMLDocument.h:72
#14 0x00007f85e2e2d889 in WebCore::DOMImplementation::createDocument(WTF::String const&, WebCore::LocalFrame*, WebCore::Settings const&, WTF::URL const&, WebCore::ProcessQualified<WTF::UUID>) (contentType=..., frame=0x7f85be1001a0, settings=..., url=..., documentIdentifier=...) at /app/webkit/Source/WebCore/dom/DOMImplementation.cpp:162
#15 0x00007f85e3bd172d in WebCore::DocumentWriter::createDocument(WTF::URL const&, WebCore::ProcessQualified<WTF::UUID>) (this=0x7f857601a0c8, url=..., documentIdentifier=...) at /app/webkit/Source/WebCore/loader/DocumentWriter.cpp:147
#16 0x00007f85e3bd18e3 in WebCore::DocumentWriter::begin(WTF::URL const&, bool, WebCore::Document*, WebCore::ProcessQualified<WTF::UUID>, WebCore::NavigationAction const*) (this=0x7f857601a0c8, urlReference=..., dispatch=false, ownerDocument=0x0, documentIdentifier=..., triggeringAction=0x7f857601a6c0) at /app/webkit/Source/WebCore/loader/DocumentWriter.cpp:159
#17 0x00007f85e3bc3a83 in WebCore::DocumentLoader::commitData(WebCore::SharedBuffer const&) (this=0x7f857601a000, data=...) at /app/webkit/Source/WebCore/loader/DocumentLoader.cpp:1264
#18 0x00007f85dfb251db in WebKit::WebLocalFrameLoaderClient::committedLoad(WebCore::DocumentLoader*, WebCore::SharedBuffer const&) (this=0x7f85be0388b0, loader=0x7f857601a000, data=...) at /app/webkit/Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:1128
#19 0x00007f85e3bc3737 in WebCore::DocumentLoader::commitLoad(WebCore::SharedBuffer const&) (this=0x7f857601a000, data=...) at /app/webkit/Source/WebCore/loader/DocumentLoader.cpp:1230
#20 0x00007f85e3bc48f5 in WebCore::DocumentLoader::dataReceived(WebCore::SharedBuffer const&) (this=0x7f857601a000, buffer=...) at /app/webkit/Source/WebCore/loader/DocumentLoader.cpp:1397
#21 0x00007f85e3bc4654 in WebCore::DocumentLoader::dataReceived(WebCore::CachedResource&, WebCore::SharedBuffer const&) (this=0x7f857601a000, resource=..., buffer=...) at /app/webkit/Source/WebCore/loader/DocumentLoader.cpp:1371
#22 0x00007f85e3d381ed in WebCore::CachedRawResource::notifyClientsDataWasReceived(WebCore::SharedBuffer const&) (this=0x7f85be21d120, buffer=...) at /app/webkit/Source/WebCore/loader/cache/CachedRawResource.cpp:144
#23 0x00007f85e3d37cc5 in WebCore::CachedRawResource::updateBuffer(WebCore::FragmentedSharedBuffer const&) (this=0x7f85be21d120, data=...) at /app/webkit/Source/WebCore/loader/cache/CachedRawResource.cpp:80
#24 0x00007f85e3cb815d in WebCore::SubresourceLoader::didReceiveBuffer(WebCore::FragmentedSharedBuffer const&, long long, WebCore::DataPayloadType) (this=0x7f85be0c6070, buffer=..., encodedDataLength=512, dataPayloadType=WebCore::DataPayloadBytes) at /app/webkit/Source/WebCore/loader/SubresourceLoader.cpp:574
#25 0x00007f85e3c96f74 in WebCore::ResourceLoader::didReceiveData(WebCore::SharedBuffer const&, long long, WebCore::DataPayloadType) (this=0x7f85be0c6070, buffer=..., encodedDataLength=512, dataPayloadType=WebCore::DataPayloadBytes) at /app/webkit/Source/WebCore/loader/ResourceLoader.cpp:606
#26 0x00007f85df9a33b9 in WebKit::WebResourceLoader::didReceiveData(IPC::SharedBufferReference&&, unsigned long) (this=0x7f85be227c00, data=..., encodedDataLength=512) at /app/webkit/Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:246
#27 0x00007f85de5a85ae in _ZZN3IPC18callMemberFunctionIN6WebKit17WebResourceLoaderES2_FvONS_21SharedBufferReferenceEmESt5tupleIJS3_mEEEEvPT_MT0_T1_OT2_ENKUlDpOT_E_clIJS3_mEEEDaSH_ (__closure=0x7ffcd6cab1e0) at /app/webkit/Source/WebKit/Platform/IPC/HandleMessage.h:136
#28 0x00007f85de5ab8c8 in _ZSt13__invoke_implIvZN3IPC18callMemberFunctionIN6WebKit17WebResourceLoaderES3_FvONS0_21SharedBufferReferenceEmESt5tupleIJS4_mEEEEvPT_MT0_T1_OT2_EUlDpOT_E_JS4_mEES9_St14__invoke_otherOSB_DpOT1_ (__f=...) at /usr/include/c++/13.2.0/bits/invoke.h:61
#29 0x00007f85de5aa47c in _ZSt8__invokeIZN3IPC18callMemberFunctionIN6WebKit17WebResourceLoaderES3_FvONS0_21SharedBufferReferenceEmESt5tupleIJS4_mEEEEvPT_MT0_T1_OT2_EUlDpOT_E_JS4_mEENSt15__invoke_resultIS9_JDpT0_EE4typeEOS9_DpOSL_ (__fn=...) at /usr/include/c++/13.2.0/bits/invoke.h:96
#30 0x00007f85de5a8615 in _ZSt12__apply_implIZN3IPC18callMemberFunctionIN6WebKit17WebResourceLoaderES3_FvONS0_21SharedBufferReferenceEmESt5tupleIJS4_mEEEEvPT_MT0_T1_OT2_EUlDpOT_E_S8_JLm0ELm1EEEDcOS9_OSB_St16integer_sequenceImJXspT1_EEE (__f=..., __t=...) at /usr/include/c++/13.2.0/tuple:2288
#31 0x00007f85de5a8656 in _ZSt5applyIZN3IPC18callMemberFunctionIN6WebKit17WebResourceLoaderES3_FvONS0_21SharedBufferReferenceEmESt5tupleIJS4_mEEEEvPT_MT0_T1_OT2_EUlDpOT_E_S8_EDcOS9_OSB_ (__f=..., __t=...) at /usr/include/c++/13.2.0/tuple:2299
#32 0x00007f85de5a86bb in IPC::callMemberFunction<WebKit::WebResourceLoader, WebKit::WebResourceLoader, void (IPC::SharedBufferReference&&, unsigned long), std::tuple<IPC::SharedBufferReference, unsigned long> >(WebKit::WebResourceLoader*, void (WebKit::WebResourceLoader::*)(IPC::SharedBufferReference&&, unsigned long), std::tuple<IPC::SharedBufferReference, unsigned long>&&) (object=0x7f85be227c00, function=(void (WebKit::WebResourceLoader::*)(WebKit::WebResourceLoader * const, IPC::SharedBufferReference &&, unsigned long)) 0x7f85df9a3014 <WebKit::WebResourceLoader::didReceiveData(IPC::SharedBufferReference&&, unsigned long)>, tuple=...) at /app/webkit/Source/WebKit/Platform/IPC/HandleMessage.h:134
#33 0x00007f85de5a6e4c in IPC::handleMessage<Messages::WebResourceLoader::DidReceiveData, WebKit::WebResourceLoader, WebKit::WebResourceLoader, void (IPC::SharedBufferReference&&, unsigned long)>(IPC::Connection&, IPC::Decoder&, WebKit::WebResourceLoader*, void (WebKit::WebResourceLoader::*)(IPC::SharedBufferReference&&, unsigned long)) (connection=..., decoder=..., object=0x7f85be227c00, function=(void (WebKit::WebResourceLoader::*)(WebKit::WebResourceLoader * const, IPC::SharedBufferReference &&, unsigned long)) 0x7f85df9a3014 <WebKit::WebResourceLoader::didReceiveData(IPC::SharedBufferReference&&, unsigned long)>) at /app/webkit/Source/WebKit/Platform/IPC/HandleMessage.h:236
#34 0x00007f85de5a62a5 in WebKit::WebResourceLoader::didReceiveWebResourceLoaderMessage(IPC::Connection&, IPC::Decoder&) (this=0x7f85be227c00, connection=..., decoder=...) at /app/webkit/WebKitBuild/GTK/Debug/DerivedSources/WebKit/WebResourceLoaderMessageReceiver.cpp:76
#35 0x00007f85df998c1f in WebKit::NetworkProcessConnection::didReceiveMessage(IPC::Connection&, IPC::Decoder&) (this=0x7f85be0383b0, connection=..., decoder=...) at /app/webkit/Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:102
#36 0x00007f85defbc5fd in IPC::Connection::dispatchMessage(IPC::Decoder&) (this=0x7f85be0603c0, decoder=...) at /app/webkit/Source/WebKit/Platform/IPC/Connection.cpp:1229
#37 0x00007f85defbc848 in IPC::Connection::dispatchMessage(WTF::UniqueRef<IPC::Decoder>) (this=0x7f85be0603c0, message=...) at /app/webkit/Source/WebKit/Platform/IPC/Connection.cpp:1277
#38 0x00007f85defbcc25 in IPC::Connection::dispatchOneIncomingMessage() (this=0x7f85be0603c0) at /app/webkit/Source/WebKit/Platform/IPC/Connection.cpp:1342
#39 0x00007f85defbc240 in operator()() const (__closure=0x7f85be2d40f8) at /app/webkit/Source/WebKit/Platform/IPC/Connection.cpp:1191
#40 0x00007f85defc39c6 in WTF::Detail::CallableWrapper<IPC::Connection::enqueueIncomingMessage(WTF::UniqueRef<IPC::Decoder>)::<lambda()>, void>::call(void) (this=0x7f85be2d40f0) at /app/webkit/WebKitBuild/GTK/Debug/WTF/Headers/wtf/Function.h:53
#41 0x00007f85d21bed45 in WTF::Function<void ()>::operator()() const (this=0x7ffcd6cab610) at /app/webkit/WebKitBuild/GTK/Debug/WTF/Headers/wtf/Function.h:82
#42 0x00007f85d38bb70b in WTF::RunLoop::performWork() (this=0x7f85be0180c0) at /app/webkit/Source/WTF/wtf/RunLoop.cpp:147
#43 0x00007f85d397844e in operator()(gpointer) const (__closure=0x0, userData=0x7f85be0180c0) at /app/webkit/Source/WTF/wtf/glib/RunLoopGLib.cpp:80
#44 0x00007f85d3978472 in _FUN(gpointer) () at /app/webkit/Source/WTF/wtf/glib/RunLoopGLib.cpp:82
#45 0x00007f85d39783e1 in operator()(GSource*, GSourceFunc, gpointer) const (__closure=0x0, source=0x5610b53c2810, callback=0x7f85d3978455 <_FUN(gpointer)>, userData=0x7f85be0180c0) at /app/webkit/Source/WTF/wtf/glib/RunLoopGLib.cpp:53
#46 0x00007f85d397842f in _FUN(GSource*, GSourceFunc, gpointer) () at /app/webkit/Source/WTF/wtf/glib/RunLoopGLib.cpp:56
#47 0x00007f85cf117d36 in g_main_dispatch (context=0x5610b5398a00) at ../glib/gmain.c:3460
#48 g_main_context_dispatch (context=0x5610b5398a00) at ../glib/gmain.c:4200
#49 0x00007f85cf1752b8 in g_main_context_iterate.isra.0 (context=0x5610b5398a00, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib/gmain.c:4276
#50 0x00007f85cf1173ff in g_main_loop_run (loop=0x5610b53c27f0) at ../glib/gmain.c:4479
#51 0x00007f85d3978aa2 in WTF::RunLoop::run() () at /app/webkit/Source/WTF/wtf/glib/RunLoopGLib.cpp:108
#52 0x00007f85dfcadc9d in WebKit::AuxiliaryProcessMainBase<WebKit::WebProcess, true>::run(int, char**) (this=0x7ffcd6cab8e0, argc=4, argv=0x7ffcd6cabaa8) at /app/webkit/Source/WebKit/Shared/AuxiliaryProcessMain.h:72
#53 0x00007f85dfcaa5d0 in WebKit::AuxiliaryProcessMain<WebKit::WebProcessMainGtk>(int, char**) (argc=4, argv=0x7ffcd6cabaa8) at /app/webkit/Source/WebKit/Shared/AuxiliaryProcessMain.h:98
#54 0x00007f85dfca251a in WebKit::WebProcessMain(int, char**) (argc=4, argv=0x7ffcd6cabaa8) at /app/webkit/Source/WebKit/WebProcess/gtk/WebProcessMainGtk.cpp:90
#55 0x00005610b3f2b969 in main(int, char**) (argc=4, argv=0x7ffcd6cabaa8) at /app/webkit/Source/WebKit/WebProcess/EntryPoint/unix/WebProcessMain.cpp:31
```

Reply:
Fixing this in https://bugs.webkit.org/show_bug.cgi?id=268344.